### PR TITLE
feat: Detectron

### PR DESCRIPTION
**Pull request by aarnphm**
We now bring detectron supports to BentoML 1.0

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

### CODE REVIEW

File: src/bentoml/_internal/frameworks/detectron.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    ...
+
+
+@t.overload
+def load_model(
+    bento_model: str | Tag | Model,
+    device_id: str | None = ...,
+    return_cfg: t.Literal[True] = ...,
+) -> tuple[nn.Module, Cf.CfgNode]:
+    ...
```
Comment (parano) [inline src/bentoml/_internal/frameworks/detectron.py:104]: what's the reason for the `t.overload` here?

File: src/bentoml/_internal/frameworks/detectron.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+if t.TYPE_CHECKING:
+    import torch.nn as nn
+
+    from .. import external_typing as ext
+    from ..models.model import ModelSignaturesType
+else:
+    nn = LazyLoader("nn", globals(), "torch.nn")
+
+
+__all__ = ["load_model", "save_model", "get_runnable", "get", "PyTorchTensorContainer"]
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py:49]: It doesn't matter right? Beacause we don't export this class anyway?

[the remark this answers is not in the document]
Comment (parano) [reply src/bentoml/_internal/frameworks/detectron.py:49]: why exposing it then?

[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py:49]: To register PyTorchTensorContainer for the datacontainer?

File: src/bentoml/_internal/frameworks/detectron.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            if torch.cuda.is_available():
+                self.device_id = "cuda"
+                torch.set_default_tensor_type("torch.cuda.FloatTensor")
+            else:
+                self.device_id = "cpu"
+
+            self.model = load_model(bento_model, device_id=self.device_id)
+
+            if not is_predictor:
+                # This predictor is a torch.nn.Module
```
Comment (sauyon) [inline src/bentoml/_internal/frameworks/detectron.py:285]:
```suggestion
                # This predictor is a torch.nn.Module
                assert isinstance(self.model, torch.nn.Module)
```

**Review by sauyon — COMMENTED** [review]
Quick pass, nothing but minor comments.

File: src/bentoml/_internal/frameworks/detectron.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+from ..runner.utils import Params
+from .common.pytorch import torch
+from .common.pytorch import inference_mode_ctx
+from .common.pytorch import PyTorchTensorContainer  # noqa # type: ignore
+
+try:
+    import detectron2.config as Cf
+    import detectron2.engine as E
+    import detectron2.modeling as M
+    import detectron2.checkpoint as C
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py:28]: done.

File: src/bentoml/_internal/frameworks/detectron.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    Example:
+
+    .. code-block:: python
+
+       import bentoml
+       # target model must be from the BentoML model store
+       model = bentoml.detectron2.get("en_reader:latest")
+    """
+    model = bentoml.models.get(tag_like)
+    if model.info.module not in (MODULE_NAME, __name__):
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py:75]: Wdu,, we did this from our FRAMEWORK_TEMPLATE_PY right?

File: src/bentoml/_internal/frameworks/detectron.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+        import bentoml
+        predictor = bentoml.detectron2.load_model('predictor:latest')
+
+        model = bentoml.detectron2.load_model('model:latest')
+    """
+    if not isinstance(bento_model, Model):
+        bento_model = get(bento_model)
+
+    if bento_model.info.module not in (MODULE_NAME, __name__):
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py:111]: same as above

File: src/bentoml/_internal/frameworks/detectron.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+        raise NotFound(
+            f"Model {bento_model.tag} was saved with module {bento_model.info.module}, not loading with {MODULE_NAME}."
+        )
+
+    cfg = bento_model.custom_objects["config"]
+    cfg.MODEL.DEVICE = device_id
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py:117]: yes

File: src/bentoml/_internal/frameworks/detectron.py [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+    model = bentoml.models.get(tag_like)
+    if model.info.module not in (MODULE_NAME, __name__):
+        raise NotFound(
+            f"Model {model.tag} was saved with module {model.info.module}, not loading with {MODULE_NAME}."
+        )
+    return model
+
+
+def load_model(
+    bento_model: str | Tag | Model, device_id: str | None = "cpu"
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py]: That was wrong. now fixed.

File: src/bentoml/_internal/frameworks/detectron.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+        model.eval()
+    else:
+        model = checkpointables.model
+        if config is not None:
+            logger.warning(
+                "config is ignored when 'checkpointables' is a 'DefaultPredictor'."
+            )
+        config = t.cast(Cf.CfgNode, checkpointables.cfg)
+
+    custom_objects["config"] = config
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/detectron.py:243]: Now thinking more about this, there is a usage for config with predictor when loading back as runner. 

I wrote a quick example about this in the documentation. So I think this is fine to save it with custom_objects.

File: src/bentoml/_internal/frameworks/detectron.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sauyon) [reply src/bentoml/_internal/frameworks/detectron.py:75]: I've forgotten, is that a backwards-compatibility thing? Seems like we can remove `__name__`.

File: src/bentoml/_internal/frameworks/detectron.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sauyon) [reply src/bentoml/_internal/frameworks/detectron.py:243]: Oh, it was for user access, I'd forgotten. Ok, let's leave this as-is, it's probably fine.

**Review by sauyon — APPROVED** [review]

**Pull request merged**